### PR TITLE
Inbox/custom reset stanza

### PR DESCRIPTION
**Pull request by NelsonVides**
As a user I want to be able to (re)set my unread count. Right now, it is reset to zero only on chat markers or messages with body, which always get broadcasted to the chat. We want to be able to set the unread count to ~any desired number (not just zero)~ _zero_, with some custom stanza that doesn’t also get broadcasted.

DONE:
- [x] The stanza looks like
```xml
<iq type='set'>
	<reset xmlns='erlang-solutions.com:xmpp:inbox:0#conversation' jid='interlocutor_jid'/>
</iq>
```
- [x] There are one test for each chat type (one2one, muclight, muc), with the following scenario: a conversation is started with a message sent, the inbox is checked by one of the receivers, to have an `unread_count` of one and that message saved, the reset stanza is sent, the result iq received is asserted, the inbox is checked again to still have the message but with an `unread_count` of zero, and it is verified that all other people did not receive this message (the reset stanza is not forwarded).
- [x] The implementation uses a new gen_iq_handler with the namespace `<<"erlang-solutions.com:xmpp:inbox:0#conversation">>`. The callback matches on `#xmlel{name = <<"reset">>}`, and checks if the jid parameters is given and is valid. If anything is not ok, answers with an error, else, does the operation and answers with the iq result.

~TODO:~
~- [x] Add test for resetting the count to numbers different than zero.~
- [x] Add tests for erroneous reset stanzas.
If you feel like these todo'es are not needed, or something else is missing, then let me know immediately 😛

### CODE REVIEW

**Comment by mongoose-im** [comment]
[7169.1](https://travis-ci.org/esl/MongooseIM/jobs/584567715) / Erlang 22.0 / small_tests / 87eab5bdbc365f04ece6316a5c0574827d1463b2
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2452/7169/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2452/7169/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-09-13_13.38.43/lib.mongooseim.logs/run.2019-09-13_13.38.43/suite.log.html)

**Comment by mongoose-im** [comment]
[7174.1](https://travis-ci.org/esl/MongooseIM/jobs/584627180) / Erlang 22.0 / small_tests / fa06b29816882fb6d639098802648983d509ab44
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2452/7174/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2452/7174/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-09-13_17.04.44/lib.mongooseim.logs/run.2019-09-13_17.04.45/suite.log.html)

**Comment by mongoose-im** [comment]
[7177.1](https://travis-ci.org/esl/MongooseIM/jobs/585142350) / Erlang 22.0 / small_tests / b03df7d40453c32ed8e3f3b1977068db9b050712
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2452/7177/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2452/7177/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-09-15_07.35.50/lib.mongooseim.logs/run.2019-09-15_07.35.50/suite.log.html)

**Comment by mongoose-im** [comment]
[7188.1](https://travis-ci.org/esl/MongooseIM/jobs/586025112) / Erlang 22.0 / small_tests / ad395a6ca5f34d7b0d88ee9f09c96448834ba947
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2452/7188/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2452/7188/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-09-17_12.02.09/lib.mongooseim.logs/run.2019-09-17_12.02.10/suite.log.html)

File: big_tests/tests/inbox_SUITE.erl [hunk 1]
```diff
@@ -762,6 +843,61 @@ non_reset_marker_should_not_affect_muclight_inbox(Config) ->
         check_inbox(Kate, [#conv{unread = 1, from = AliceRoomJid,
                                  to = KateJid, content = Msg}])
       end).
+
+groupchat_markers_one_reset_reset_stanza(Config) ->
+    escalus:story(Config, [{alice, 1}, {bob, 1}, {kate, 1}], fun(Alice, Bob, Kate) ->
+        % %% WITH
+        AliceJid = inbox_helper:to_bare_lower(Alice),
```
[the remark this answers is not in the document]
Comment (NelsonVides) [reply big_tests/tests/inbox_SUITE.erl]: I basically tried to follow the style already present on the SUITE, not importing or unimporting anything that wasn't already there 🤷‍♂

File: big_tests/tests/inbox_SUITE.erl [hunk 2]
```diff
@@ -762,6 +843,61 @@ non_reset_marker_should_not_affect_muclight_inbox(Config) ->
...
+        BobJid = inbox_helper:to_bare_lower(Bob),
+        KateJid = inbox_helper:to_bare_lower(Kate),
+        RoomJid = room_bin_jid(?ROOM_MARKERS_RESET),
+        AliceRoomJid = <<RoomJid/binary,"/", AliceJid/binary>>,
+        % %% WHEN A MESSAGE IS SENT
+        MsgStanza = escalus_stanza:set_id(
+          escalus_stanza:groupchat_to(RoomJid, <<"marker time!">>), <<"some_ID">>),
+        escalus:send(Alice, MsgStanza),
+        R0 = escalus:wait_for_stanza(Alice),
```
[the remark this answers is not in the document]
Comment (NelsonVides) [reply big_tests/tests/inbox_SUITE.erl]: Ah, that was already existing! 😄

File: big_tests/tests/inbox_SUITE.erl [hunk 3]
```diff
@@ -1122,6 +1258,56 @@ non_reset_marker_should_not_affect_muc_inbox(Config) ->
...
+    escalus:story(Config, [{alice, 1}, {bob, 1}, {kate, 1}], fun(Alice, Bob, Kate) ->
+        % %% WITH
+        Users = [Alice, Bob, Kate],
+        Msg = <<"Hi Room!">>,
+        Id = <<"MyID">>,
+        Room = ?config(room, Config),
+        RoomAddr = muc_helper:room_address(Room),
+        % %% PROVIDED
+        inbox_helper:enter_room(Room, Users),
+        inbox_helper:make_members(Room, Alice, Users -- [Alice]),
```
[the remark this answers is not in the document]
Comment (NelsonVides) [reply big_tests/tests/inbox_SUITE.erl]: As said, I pretty much followed the style already present in the SUITE so, it was already done like that everywhere. Not saying it's a better way, it's just an unfortunate consistency 🤷‍♂

**Comment by mongoose-im** [comment]
[7189.1](https://travis-ci.org/esl/MongooseIM/jobs/586122659) / Erlang 22.0 / small_tests / 891d905a9023a2fbc330804712939a40036e465c
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2452/7189/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2452/7189/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-09-17_15.44.23/lib.mongooseim.logs/run.2019-09-17_15.44.24/suite.log.html)

File: big_tests/tests/inbox_SUITE.erl [hunk 4]
```diff
@@ -762,6 +812,39 @@ non_reset_marker_should_not_affect_muclight_inbox(Config) ->
         check_inbox(Kate, [#conv{unread = 1, from = AliceRoomJid,
                                  to = KateJid, content = Msg}])
       end).
+
+groupchat_markers_one_reset_reset_stanza(Config) ->
```
Comment (fenek) [inline big_tests/tests/inbox_SUITE.erl]: The name of this case is completely unclear.

File: big_tests/tests/inbox_helper.erl [hunk 5]
```diff
@@ -251,6 +255,40 @@ make_inbox_stanza(GetParams, Verify) ->
...
+        ResetStanza = make_reset_inbox_stanza(To),
+        escalus:send(From, ResetStanza),
+        Result = escalus:wait_for_stanza(From),
+        ?assert(escalus_pred:is_iq_result(ResetStanza, Result)),
+        ?assertNotEqual(undefined, exml_query:path(Result, [{element_with_ns, <<"reset">>,
+                                                             ?NS_ESL_INBOX_CONVERSATION}])).
+
+-spec make_reset_inbox_stanza(
+        undefined | jid:jid() | escalus:client() | atom() | string() | binary()) -> exml:element().
+make_reset_inbox_stanza(InterlocutorJid) when is_binary(InterlocutorJid) ->
```
Comment (fenek) [inline big_tests/tests/inbox_helper.erl]: Are all these cases for `InterlocutorJid` types even used? I'm pretty sure that `escalus:client()`, `undefined` and `binary()` are sufficient.

File: big_tests/tests/inbox_helper.erl [hunk 6]
```diff
@@ -574,13 +612,23 @@ send_and_mark_msg(From, To) ->
     escalus:send(To, ChatMarker),
     Msg.
 
+assert_has_no_stanzas(UsersList) when is_list(UsersList) ->
+    lists:foreach(fun(User) -> ?assertNot(escalus_client:has_stanzas(User)) end, UsersList);
+assert_has_no_stanzas(User) ->
+    ?assertNot(escalus_client:has_stanzas(User)).
+
+assert_invalid_reset_inbox_form(From, To, Field, Value) ->
```
Comment (fenek) [inline big_tests/tests/inbox_helper.erl]: We are not sending any forms in custom reset IQ actually.

File: doc/modules/mod_inbox.md [hunk 7]
```diff
@@ -79,6 +79,26 @@ Server:
 </iq>
 ```
 
+### Reseting inbox
+
+An inbox can be reset at will using the stanza
```
Comment (goddammit) [inline doc/modules/mod_inbox.md]: You can reset the inbox with the following stanza:

File: doc/modules/mod_inbox.md [hunk 8]
```diff
@@ -79,6 +79,26 @@ Server:
...
+
+```xml
+<iq type='set'>
+    <reset xmlns='erlang-solutions.com:xmpp:inbox:0#conversation' jid='interlocutor_bare_jid'/>
+</iq>
+```
+
+Where `jid` is the bare jid of the interlocutor whose inbox wants to be reset.
```
Comment (goddammit) [inline doc/modules/mod_inbox.md]: Here  `jid` is the bare jid of the user whose inbox we want to reset.

File: doc/modules/mod_inbox.md [hunk 9]
```diff
@@ -79,6 +79,26 @@ Server:
...
+This action does not change the last message stored in inbox; that is, nor this
```
Comment (goddammit) [inline doc/modules/mod_inbox.md]: This action does not change the last message stored in inbox; meaning that neither this

File: doc/modules/mod_inbox.md [hunk 10]
```diff
@@ -79,6 +79,26 @@ Server:
...
+stanza nor anything given within will be stored; only the inbox `unread_count`
```
Comment (goddammit) [inline doc/modules/mod_inbox.md]: stanza nor anything given within will be stored; the only change is the inbox `unread_count`

File: doc/modules/mod_inbox.md [hunk 11]
```diff
@@ -79,6 +79,26 @@ Server:
...
+changes to zero.
```
Comment (goddammit) [inline doc/modules/mod_inbox.md]: is set to zero.

File: doc/modules/mod_inbox.md [hunk 12]
```diff
@@ -79,6 +79,26 @@ Server:
...
+
+Resetting the inbox count by this stanza will also skip the forwarding of
```
Comment (goddammit) [inline doc/modules/mod_inbox.md]: Resetting the inbox count  will also skip the forwarding of

File: doc/modules/mod_inbox.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (NelsonVides) [reply doc/modules/mod_inbox.md]: Which jid, the one inside of the code or the one inside of the docs, or both?

File: doc/modules/mod_inbox.md [hunk 13]
```diff
@@ -1,7 +1,7 @@
 ### Module Description
 
-`Inbox` is an experimental feature implemented as a few seperate modules.
-To use it, enable mod_inbox in the config file.
+`Inbox` is an experimental feature implemented as a few separate modules.
+To use it, enable mod\_inbox in the config file.
```
[the remark this answers is not in the document]
Comment (NelsonVides) [reply doc/modules/mod_inbox.md]: No, technically markdown requires escaping the character. `_` means italicising. If GitHub solves the problem that doesn't make it any less correct. The problem about markdown is that the specifications are not really standardised and all parsers implementers have their own slight variations. Some, require escaping these things.

File: big_tests/tests/inbox_helper.erl (hunk 5 above)
Comment (NelsonVides) [reply big_tests/tests/inbox_helper.erl]: I though of leaving the application flexible so to not have to worry about what is being passed, but whatever, I'm deleting it.

**Comment by mongoose-im** [comment]
[7191.1](https://travis-ci.org/esl/MongooseIM/jobs/586453886) / Erlang 22.0 / small_tests / 558f1d0e6c26edf0daaed386f27903680f222e15
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2452/7191/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2452/7191/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-09-18_09.28.29/lib.mongooseim.logs/run.2019-09-18_09.28.30/suite.log.html)

File: src/inbox/mod_inbox.erl [hunk 14]
```diff
@@ -157,6 +165,32 @@ process_iq(From, _To, Acc, #iq{type = set, id = QueryId, sub_el = QueryEl} = IQ)
...
+maybe_process_reset_stanza(From, Acc, IQ, ResetStanza) ->
+    case reset_stanza_extract_interlocutor_jid(ResetStanza) of
+        {error, Msg} ->
+            {Acc, IQ#iq{type = error, sub_el = [mongoose_xmpp_errors:bad_request(<<"en">>, Msg)]}};
+        InterlocutorJID ->
+            process_reset_stanza(From, Acc, IQ, ResetStanza, InterlocutorJID)
+    end.
+
+process_reset_stanza(From, Acc, IQ, _ResetStanza, InterlocutorJID) ->
+    ok = mod_inbox_utils:reset_unread_count(From, InterlocutorJID, 0),
```
[the remark this answers is not in the document]
Comment (erszcz) [reply src/inbox/mod_inbox.erl]: It's easy to add anyway if need be.

**Comment by mongoose-im** [comment]
[7193.1](https://travis-ci.org/esl/MongooseIM/jobs/586547865) / Erlang 22.0 / small_tests / 63f38c93ca67ee804b560428a987654101e26851
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2452/7193/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2452/7193/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-09-18_13.50.20/lib.mongooseim.logs/run.2019-09-18_13.50.21/suite.log.html)

**Comment by NelsonVides** [comment]
> Code looks good to me. I'm a bit conservative when it comes to CI status before merge so I'd like to wait until travis or CircleCi is green for this PR.

All green: https://travis-ci.org/esl/MongooseIM/builds/586547864 🙂
